rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala

```diff
@@ -526,6 +530,27 @@ object ProcNormalizeMatcher {
         ProcVisitOutputs(input.par.prepend(foldedNew), bodyResult.knownFree)
       }
 
+      case b: PBundle =>
+        val targetResult = normalizeMatch(b.proc_, input.copy(par = VectorPar()))
+        if (targetResult.par.wildcard || targetResult.par.freeCount > 0) {
+          val errMsg = {
+            def at(variable: String, l: Int, col: Int): String =
+              s"$variable line: $l, column: $col"
+            val wildcardsPositions = targetResult.knownFree.wildcards.map {
+              case (l, col) => at("", l, col)
+            }
+            val freeVarsPositions = targetResult.knownFree.env.map {
+              case (n, (_, _, line, col)) => at(s"`$n`", line, col)
+            }
+            wildcardsPositions.mkString(" Wildcards at positions: ", ", ", ".") ++
+              freeVarsPositions.mkString(" Free variables at positions: ", ", ", ".")
+          }
+          throw UnexpectedBundleContent(
+            s"Bundle's content shouldn't have free variables or wildcards.$errMsg")
```

Comment:
So essentially we are relying on the normalizer to enforce the "bundling"? What if I bypass the normalizer and just send code straight to the interpreter or will that not be possible in our final design?

Reply:
No, that won't be possible in the final design, but also:
The security of bundling doesn't come from there, it comes from the matcher, which won't match free variables inside of a bundle anyway.